lib/temporal/client.rb

```diff
@@ -0,0 +1,382 @@
+require 'json'
+require 'securerandom'
+require 'socket'
+require 'temporal/errors'
+require 'temporal/converter'
+require 'temporal/interceptor/client'
+require 'temporal/interceptor/chain'
+require 'temporal/workflow/handle'
+require 'temporal/workflow/execution_info'
+require 'temporal/workflow/id_reuse_policy'
+require 'temporal/workflow/query_reject_condition'
+
+module Temporal
+  class Client
+    attr_reader :namespace, :identity
+
+    # TODO: More argument to follow for converters, codecs, etc
+    def initialize(connection, namespace, interceptors: [])
+      @connection = connection
+      @namespace = namespace
+      @identity = "#{Process.pid}@#{Socket.gethostname}"
+      @interceptor_chain = Interceptor::Chain.new(interceptors)
+      @converter = Converter.new
+    end
+
+    def start_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      execution_timeout: nil,
+      run_timeout: nil,
+      task_timeout: nil,
+      id_reuse_policy: Workflow::IDReusePolicy::ALLOW_DUPLICATE,
+      retry_policy: nil,
+      cron_schedule: '',
+      memo: nil,
+      search_attributes: nil,
+      headers: {},
+      start_signal: nil,
+      start_signal_args: []
+    )
+      retry_policy&.validate!
+
+      input = Interceptor::Client::StartWorkflowInput.new(
+        workflow: workflow,
+        args: args,
+        id: id,
+        task_queue: task_queue,
+        execution_timeout: execution_timeout,
+        run_timeout: run_timeout,
+        task_timeout: task_timeout,
+        id_reuse_policy: id_reuse_policy,
+        retry_policy: retry_policy,
+        cron_schedule: cron_schedule,
+        memo: memo,
+        search_attributes: search_attributes,
+        headers: headers,
+        start_signal: start_signal,
+        start_signal_args: start_signal_args,
+      )
+
+      interceptor_chain.invoke(:start_workflow, input) do |i|
+        handle_start_workflow(i)
+      end
+    end
+
+    def describe_workflow(id, run_id)
+      input = Interceptor::Client::DescribeWorkflowInput.new(
+        id: id,
+        run_id: run_id,
+      )
+
+      interceptor_chain.invoke(:describe_workflow, input) do |i|
+        handle_describe_workflow(i)
+      end
+    end
+
+    def query_workflow(
```

Comment:
I believe that this help with the readability (which in my view Python and TS SDKs lack significantly). You have 2 interfaces — the public API and the private implementation that accepts combined `input`s. I'm also not sure why the cross SDK similarity is relevant here as long as the public interface is similar and behaviour doesn't change.

Reply:
It's not that the similarity is relevant, it's that the reasons for not having this at the top level _and_ in the handle are probably cross-SDK reasons and if we consider those reasons good enough, they are probably good enough for more than Ruby. And if we don't consider those reasons good enough, they are probably not good enough for any of them.

Reply:
I agree that it's nice to have these functions exist on the handle rather than both places - they always need to operate on a workflow instance, so segregating:

* reduces top level API surface of client, and moves some things to handle, providing better organization
* reduces number of arguments to these methods (run / workflow id are built-in to the handle)

Reply:
Ah, right I see what you mean. I can move the implementations into a separate class. To me a Handle is not something that should implement anything, it's just an access convenience thing. If we want to hide these extra endpoints from the Client — I'm fine with moving most of the implementation to a private class that both Client and Handle will interact with. It's the fact that the implementation of the same level logic is spread across multiple places that bothers me, hence implementing all of it in the Client and not the handle

Reply:
Why not just implement it all in the (final) interceptor and not in the client or handle?

Reply:
Because not all the Client/Handle methods are intercepted, namely the `result`. I don't see the reason why it has to be special and be implemented elsewhere. This means that the handle now has to be coupled to a specific service, converters and protobuf definitions. And combining these all together is not really solving anything (just making dependencies a bit less obvious). A good litmus test here would be unit tests — they would expose the inefficiencies of this design